src/main/java/Links.java

```diff
@@ -1,5 +1,139 @@
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParser;
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.select.Elements;
+
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
 interface Links extends Iterable<URL> {
+    Iterator<URL> iterator();
+    String toString();
+
     class HTML implements Links {
+        private String inputURL;
+        private ArrayList<URL> links;
+        private HashMap<String, ArrayList<URL>> deadLinks;
+        public String errorStatus;
+
+        private boolean isValidLink(String url) {
+            String urlRegex = "^(http|https)://[-a-zA-Z0-9+&@#/%?=~_|,!:.;]*[-a-zA-Z0-9+@#/%=&_|]";
```

Comment:
_Regular expressions are double-edged weapons: if you have a problem and solve it with regex, you now have two problems._

Point here is – nowadays URLs may contain non-ASCII characters and be a bit more complex than one initially considers, so it may be easier/faster to rely on a third-party validation (unless it's necessary to control the validation process happening behind the scenes). Some of the options here are to instantiate `java.net.URL` or use Apache Commons `UrlValidator`.